dev-docs/RFCs/vNext/icon-layer-dynamic-image-sources.md

```diff
@@ -0,0 +1,144 @@
+  -  when constructing a `IconLayer`, `iconMppping` and `iconAtlas` are not needed. 
+  - The expected `object` should include:
+    * `url`: url to fetch the icon
+    * `height`: height of icon
+    * `width`: width of icon
+    * `anchorX`: horizontal position of icon anchor. Default: half width.
+    * `anchorY`: vertical position of icon anchor. Default: half height.
+    * `mask`: whether icon is treated as a transparency mask.
+       - If `true`, user defined color is applied.
+       - If `false`, pixel color from the image is applied. User still can specify the opacity through getColor.
+      Default: `false`
+ 
+## Code examples 
+
+- `getIcon` return a string, also require to `IconLayer`
+```js
+import DeckGL, {IconLayer} from 'deck.gl';
+
+const ICON_MAPPING = {
+  marker: {x: 0, y: 0, width: 32, height: 32, mask: true}
+};
+
+const App = ({data, viewport}) => {
+
+  /**
+   * Data format:
+   * [
+   *   {name: 'Colma (COLM)', address: '365 D Street, Colma CA 94014', exits: 4214, coordinates: [-122.466233, 37.684638]},
+   *   ...
+   * ]
+   */
+  const layer = new IconLayer({
+    id: 'icon-layer',
+    data,
+    pickable: true,
+
+    // `iconAtlas` and `iconMapping` are required
+    iconAtlas: 'images/icon-atlas.png',
+    iconMapping: {
+      marker: {
+        x: 0,
+        y: 0,
+        width: 128,
+        height: 128,
+        anchorY: 128,
+        mask: true
+      }
+    },
+    // return a string
+    getIcon: d => 'marker',
+    
+    sizeScale: 15,
+    getPosition: d => d.coordinates,
+    getSize: d => 5,
+    getColor: d => [Math.sqrt(d.exits), 140, 0]
+  });
+
+  return (<DeckGL {...viewport} layers={[layer]} />);
+};
+```
+
+- `getIcon` return an object
+```js
+import DeckGL, {IconLayer} from 'deck.gl';
+
+const ICON_MAPPING = {
+  marker: {x: 0, y: 0, width: 32, height: 32, mask: true}
+};
+
+const App = ({data, viewport}) => {
+
+  /**
+   * Data format:
+   * [
+   *   {name: 'Colma (COLM)', avatar_url: 'https://data/colma_avatar.png', address: '365 D Street, Colma CA 94014', exits: 4214, coordinates: [-122.466233, 37.684638]},
+   *   ...
+   * ]
+   */
+  const layer = new IconLayer({
+    id: 'icon-layer',
+    data,
+    pickable: true,
+    sizeScale: 15,
+    getPosition: d => d.coordinates,
+    
+    // `iconAtlas` and `iconMapping` are not needed
+    // return an object 
+    getIcon: d => ({
+      url: d.avatar_url,
+      width: 128,
+      height: 128,
+      anchorY: 128,
+      mask: true
+    }),
+    
+    getSize: d => 5,
+    getColor: d => [Math.sqrt(d.exits), 140, 0]
+  });
+
+  return (<DeckGL {...viewport} layers={[layer]} />);
+};
+```
+
+## Cost and Impact
+
+The proposal will require the following changes:
+- `IconManager` class: add a new class to help fetch icons and manage `iconMapping` and `iconAltas` from 
+dynamically fetched icons. 
+- `IconLayer` needs update to handle both pre-defined sprite image and dynamic fetched icons.
+
+If implemented:
```

Comment:
Analyze the performance impact of using prepacked icon atlas vs. auto packing. The old approach is still way more efficient, and should be encouraged in all use cases where it's possible. It is also important to clarify this when the layer documentation is updated.